Quotient/room.cpp

```diff
@@ -2878,6 +2880,51 @@ inline bool isEditing(const RoomEventPtr& ep)
                      false);
 }
 
+void Room::Private::processRedactionsAndEdits(RoomEvents& events)
+{
+    // Pre-process redactions and edits so that events that get
+    // redacted/replaced in the same batch landed in the timeline already
+    // treated.
+    // NB: We have to store redacting/replacing events to the timeline too -
+    // see #220.
+    auto it = std::find_if(events.begin(), events.end(), isEditing);
+    for (const auto& eptr : RoomEventsRange(it, events.end())) {
+        if (auto* r = eventCast<RedactionEvent>(eptr)) {
+            // Try to find the target in the timeline, then in the batch.
+            if (processRedaction(*r))
+                continue;
+            if (auto targetIt = std::find_if(events.begin(), events.end(),
+                    [id = r->redactedEvent()](const RoomEventPtr& ep) {
+                        return ep->id() == id;
+                    }); targetIt != events.end())
+                *targetIt = makeRedacted(**targetIt, *r);
+            else
+                qCDebug(STATE)
+                    << "Redaction" << r->id() << "ignored: target event"
+                    << r->redactedEvent() << "is not found";
+            // If the target event comes later, it comes already redacted.
+        }
+        if (auto* msg = eventCast<RoomMessageEvent>(eptr);
+                msg && !msg->replacedEvent().isEmpty()) {
+            if (processReplacement(*msg))
+                continue;
+            auto targetIt = std::find_if(events.begin(), events.end(),
+                    [id = msg->replacedEvent()](const RoomEventPtr& ep) {
+                        return ep->id() == id;
+                    });
+            if (targetIt != events.end())
+                *targetIt = makeReplaced(**targetIt, *msg);
```

Comment:
nit: Since we're moving this block, can we use std::exchange. Ditto for the first branch too